### DISCUSSION TITLE
How to use API access token from inside a pod?

### DISCUSSION THREAD

**Discussion by abravalheri** [argoproj/argo-workflows#13158] [Q&A] [answered]
Hello, I am having some trouble to understand how to trigger workflows from a script/program running inside of a pod on the same cluster.

I went through the docs in https://argo-workflows.readthedocs.io/en/stable/access-token/, and I can only find examples using the `kubectl` command to acquire the token, etc... Is there any example of how to achieve that without access to `kubectl`, or how to trigger workflows from a script/program running inside a pod?

**Comment by agilgur5** [comment, marked as answer]
Is there a reason you need API access if you're using a Pod in the same cluster? You can talk to k8s directly from the Pod, no need to go through the API (in most cases at least). The token is the credentials of a Service Account; if you're already in the cluster, you can use the SA directly.

To answer the question though, the token is a standard [k8s `Secret`](https://kubernetes.io/docs/tasks/inject-data-application/distribute-credentials-secure/); you can mount it as a volume or use it as an env var like any other k8s `Secret`

**Comment by abravalheri** [comment]
> Is there a reason you need API access if you're using a Pod in the same cluster?

Well, I suppose so? 😅
Imagine the following: there is an application deployed inside of a pod on a cluster.
From time to time, users make a request to this application that, in turn, translates to a workflow being triggered. The pod running the application needs to be able to access Argo's API so that it can submit the workflow, right?

I suppose we need the access credentials in order to be able to do that...
But please let me know if you see any other way of doing it.

> To answer the question though, the token is a standard [k8s Secret](https://kubernetes.io/docs/tasks/inject-data-application/distribute-credentials-secure/); you can mount it as a volume or use it as an env var like any other k8s Secret

Thank you very much!

**Reply by agilgur5** [reply]
> The pod running the application needs to be able to access Argo's API so that it can submit the workflow, right?

Not necessarily, you can submit it directly to the k8s api-server. For instance, with `kubectl` or a k8s SDK. Or  via the `argo` CLI in [k8s mode](https://argo-workflows.readthedocs.io/en/latest/cli/argo/#synopsis).

The Argo API is an intermediary to k8s and is mostly an optional component (except for certain features like retrieving [archived](https://argo-workflows.readthedocs.io/en/latest/workflow-archive/) or [offloaded](https://argo-workflows.readthedocs.io/en/latest/offloading-large-workflows/) workflows)

**Reply by abravalheri** [reply]
Thank you @agilgur5.

I see now, I did not know that the workflow would work as a custom CRD with vanilla kubernetes. Thank you very much for the information. That is very useful.
